jwst/undersampling_correction/undersampling_correction.py

```diff
@@ -0,0 +1,107 @@
+#  Module for undersampling correction
+#
+import logging
+import numpy as np
+
+from stdatamodels.jwst.datamodels import dqflags
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+def undersampling_correction(input_model, signal_threshold):
+    """
+    Correct for undersampling.
+
+    Parameters
+    ----------
+    input_model : `~jwst.datamodels.RampModel`
+        The input science data to be corrected
+
+    signal_threshold : float
+        Science value above which a group will be flagged as UNDERSAMP
+
+    Returns
+    -------
+    output_model : `~jwst.datamodels.RampModel`
+        Data model with undersampling_correction applied; add UNDERSAMP flag
+        to groups exceeding signal_threshold
+    """
+    data = input_model.data
+    gdq = input_model.groupdq
+
+    # Create the output model as a copy of the input
+    output_model = input_model.copy()
+
+    log.info('Using signal_threshold: %.2f', signal_threshold)
+
+    gdq_new = flag_pixels(data, gdq, signal_threshold)
+
+    # Save the flags in the output GROUPDQ array
+    output_model.groupdq = gdq_new
+
+    return output_model
+
+
+def flag_pixels(data, gdq, signal_threshold):
+    """
+    Flag first group in each ramp that exceeds signal_threshold as UNDERSAMP and DO_NOT_USE,
+    skipping groups already flagged as DO_NOT_USE; then flag all subsequent groups in the ramp.
+
+    Parameters
+    ----------
+    data : float, 4D array
+        science array
+
+    gdq : int, 4D array
+        group dq array
+
+    signal_threshold : float
+        Science value above which a group will be flagged as UNDERSAMP and DO_NOT_USE
+
+    Returns
+    -------
+    gdq : int, 4D array
+        updated group dq array
+    """
+    n_ints = gdq.shape[0]
+    n_grps = gdq.shape[1]
+    n_rows = gdq.shape[2]
+    n_cols = gdq.shape[3]
```

Comment:
This is probably a trivial performance hit but is really not necessary to access `shape` 4 times as python allows unpacking of tuple values in one call
```
n_ints, n_grps, n_rows, n_cols = gdq.shape
```